pkg/apis/kops/validation/legacy.go

```diff
@@ -55,41 +55,54 @@ func ValidateCluster(c *kops.Cluster, strict bool) field.ErrorList {
 	requiresSubnets := true
 	requiresNetworkCIDR := true
 	requiresSubnetCIDR := true
-	switch kops.CloudProviderID(c.Spec.CloudProvider) {
-	case "":
-		allErrs = append(allErrs, field.Required(fieldSpec.Child("cloudProvider"), ""))
+
+	optionTaken := false
+	if c.Spec.CloudProvider.AWS != nil {
+		optionTaken = true
+
+	}
+	if c.Spec.CloudProvider.Azure != nil {
+		if optionTaken {
+			allErrs = append(allErrs, field.Forbidden(fieldSpec.Child("azure"), "only one cloudProvider option permitted"))
+		}
+		optionTaken = true
+	}
+	if c.Spec.CloudProvider.DO != nil {
+		if optionTaken {
+			allErrs = append(allErrs, field.Forbidden(fieldSpec.Child("do"), "only one cloudProvider option permitted"))
+		}
+		optionTaken = true
 		requiresSubnets = false
 		requiresSubnetCIDR = false
 		requiresNetworkCIDR = false
-
-	case kops.CloudProviderGCE:
+		if c.Spec.NetworkCIDR != "" {
+			allErrs = append(allErrs, field.Forbidden(fieldSpec.Child("networkCIDR"), "networkCIDR should not be set on DigitalOcean"))
+		}
+	}
+	if c.Spec.CloudProvider.GCE != nil {
+		if optionTaken {
+			allErrs = append(allErrs, field.Forbidden(fieldSpec.Child("gce"), "only one cloudProvider option permitted"))
+		}
+		optionTaken = true
 		requiresNetworkCIDR = false
+		requiresSubnetCIDR = false
 		if c.Spec.NetworkCIDR != "" {
 			allErrs = append(allErrs, field.Forbidden(fieldSpec.Child("networkCIDR"), "networkCIDR should not be set on GCE"))
 		}
+	}
+	if c.Spec.CloudProvider.Openstack != nil {
+		if optionTaken {
+			allErrs = append(allErrs, field.Forbidden(fieldSpec.Child("openstack"), "only one cloudProvider option permitted"))
+		}
+		optionTaken = true
+		requiresNetworkCIDR = false
 		requiresSubnetCIDR = false
-
-	case kops.CloudProviderDO:
+	}
+	if !optionTaken {
+		allErrs = append(allErrs, field.Required(fieldSpec.Child("cloudProvider"), ""))
```

Comment:
Is there a reason to set requiresSubnets /  requiresSubnetCIDR / requiresNetworkCIDR  here?  Changing the order makes the change a harder to review (I'll check the individual PRs)

Reply:
I figured alphabetizing would make it easier to maintain going forward.